core/src/main/java/io/keen/client/java/KeenClient.java

```diff
@@ -1360,14 +1362,22 @@ private void validateEventValue(Object value, int depth) {
      * @param event           The event to publish.
      * @return The response from the server.
      * @throws IOException If there was an error communicating with the server.
+     * @throws URISyntaxException if you write illegal character to encode.
      */
     private String publish(KeenProject project, String eventCollection,
                            Map<String, Object> event) throws IOException {
+        URL url = null;
         // just using basic JDK HTTP library
-        String urlString = String.format(Locale.US, "%s/%s/projects/%s/events/%s", getBaseUrl(),
-                KeenConstants.API_VERSION, project.getProjectId(), eventCollection);
-        URL url = new URL(urlString);
+        try {
+            eventCollection = new URI(null, null, eventCollection, null).getRawPath();
+            String urlString = String.format(Locale.US, "%s/%s/projects/%s/events/%s", getBaseUrl(),
+                    KeenConstants.API_VERSION, project.getProjectId(), eventCollection);
+            url = new URL(urlString);
+        } catch (URISyntaxException e) {
+            KeenLogging.log("Event collection name has invalid character to encode", e);
```

Comment:
and it means that you'd like to publishObject with null URL what will fail...

I'd suggest extracting a new method (sth like createURL(...)) that returns Optional type and calling it here. When it returns empty optional then you should throw a IllegalStateException with relevant message instead of calling publishObject with null IRL

Reply:
*URL